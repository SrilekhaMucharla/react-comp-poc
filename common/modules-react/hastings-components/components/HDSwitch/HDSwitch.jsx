import React, { useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const HDSwitch = ({
    id,
    values,
    value,
    theme,
    onChange,
    path,
    name
}) => {
    const [val, setVal] = useState(value);

    const handleChange = (event) => {
        const { checked } = event.target;
        const newValue = checked ? values[1].value : values[0].value;
        setVal(newValue);
        onChange(
            {
                target: {
                    value: newValue,
                    name: name,
                    label: checked ? values[1].name : values[0].name,
                    getAttribute: (attr) => (attr === 'path' ? path : null)
                }
            }
        );
    };

    const keyId = id || 'hd-custom-switch';
    const isChecked = val === values[1].value;

    return (
        <div className={classNames('hd-switch', `switch--${theme}`)}>
            <div className={classNames('custom-control', 'custom-switch')}>
                <input
                    name={name}
                    type="checkbox"
                    className="custom-control-input"
                    id={keyId}
                    onChange={handleChange}
                    checked={isChecked} />
                <label
                    htmlFor={keyId}
                    className={classNames('custom-control-label-left', !isChecked && 'active')}
                >
                    {values[0].name}
                </label>
                <label
                    htmlFor={keyId}
                    className={classNames('custom-control-label', 'custom-control-label-right', isChecked && 'active')}
                >
                    {values[1].name}
                </label>
            </div>
        </div>
    );
};

HDSwitch.propTypes = {
    id: PropTypes.string,
    theme: PropTypes.oneOf(['dark', 'light']),
    values: PropTypes.arrayOf(PropTypes.shape({
        value: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
    })).isRequired,
    onChange: PropTypes.func,
    value: PropTypes.string.isRequired,
    path: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired

};

HDSwitch.defaultProps = {
    id: null,
    theme: 'dark',
    onChange: () => {},
};

HDSwitch.typeName = 'HDSwitch';

export default HDSwitch;
