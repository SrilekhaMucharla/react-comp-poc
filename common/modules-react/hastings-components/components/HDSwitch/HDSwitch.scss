@mixin switch($theme: 'dark') {
  $active: $white-fill;
  $non-active: $info-color;
  $hover-active: $blue-light;
  $hover-non-active: $white-fill;
  @if $theme == 'light' {
    $active: $blue-dark-4;
    $non-active: $blue-light;
    $hover-active: $info-color;
    $hover-non-active: $blue-dark-4;
  }

    .custom-switch {
      align-items: baseline;
      display: inline-flex;
      padding-left: 0;

      .custom-control-input {
        z-index: -1;
    
        &:hover ~ .custom-control-label {

          &::after {
            transform: translateX(12px);
          }

          &::before{
            border: solid 2px $blue-color;
            background-color: $link-color;
            box-shadow: none; 
          }

          &-right, &-left {
            color: $hover-non-active;
            &.active {
              color: $hover-active;
            }
          }
        }
      
        &:checked ~ .custom-control-label::after {
          transform: translateX(42px);
        }

        &:checked:hover ~ .custom-control-label::after {
          transform: translateX(30px);
        }

        &:checked ~ .custom-control-label::before {
          background-color: $link-color;
        }

        &:focus ~ .custom-control-label::before {
          box-shadow: none; 
        } 
      }

      .custom-control-label{

        &-right, &-left {
          //not active label
          font-weight: bold;
          color: $non-active;
          &.active {
            // active label
            font-weight: bold;
            color: $active;
          }
        }

        &-right {
          padding-left: 3.25rem;
          padding-top: 6px;
        }
        
        &-left {
          position: relative;
          padding-right: 3rem;
          display: inline;
        }
      
        &::before {
          left: -2.25rem;
          width: 78px;
          height: 36px;
          pointer-events: all;
          border-radius: 100px;
          cursor: pointer;
          border: solid 2px $link-color;
          background-color: $link-color;
          outline: none;
          box-shadow: none;
          margin-top: -4px;
        }

        &::after {
          width: 32px;
          height: 32px;
          border-radius: 100px;
          cursor: pointer;
          background-color: $white-fill;
          margin-top: -4px;
        }
      }
    }
  
}

.switch--dark {
  @include switch('dark');
}

.switch--light {
  @include switch('light');
}



