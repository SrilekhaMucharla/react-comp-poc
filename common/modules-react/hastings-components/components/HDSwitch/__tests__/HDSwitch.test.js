import React from 'react';
import { shallow } from 'enzyme';
import HDSwitch from '../HDSwitch';

describe('HDSwitch', () => {
    // given
    const onChange = jest.fn();
    const values = [
        {
            name: 'value 1',
            value: '1'
        },
        {
            name: 'value 2',
            value: '2'
        }
    ];
    const { value } = values[1];
    const name = 'name';
    const path = 'path';

    const getWrapper = (props = {}) => shallow(
        <HDSwitch
            path={path}
            name={name}
            value={value}
            onChange={onChange}
            values={values}
            {...props} />
    );

    it('should render correctly and match the snapshot', () => {
        // given
        const wrapper = getWrapper();
        // then
        expect(wrapper).toMatchSnapshot();
    });

    it('should contain two labels', () => {
        // given
        const wrapper = getWrapper();
        // then
        expect(wrapper.find('label')).toHaveLength(2);
    });

    it('should contain input', () => {
        // given
        const wrapper = getWrapper();
        // then
        expect(wrapper.find('input')).toHaveLength(1);
    });

    it('should call onChange when switch is clicked off', () => {
        // given
        const defaultChecked = false;
        const wrapper = getWrapper({
            defaultChecked
        });
        const customEvent = {
            target: {
                value: values[1].value,
                name: name,
                getAttribute: (attr) => (attr === 'path' ? path : null)
            }
        };

        // when
        wrapper.find('.custom-control-input').simulate('change', { target: { checked: !defaultChecked } });
        // then
        expect(onChange).toHaveBeenCalledWith(expect.objectContaining({
            target: expect.objectContaining({
                value: customEvent.target.value,
                name: customEvent.target.name
            })
        }));
        expect(wrapper.find('.custom-control-label-right').hasClass('active')).toBeTruthy();
    });

    it('should call onChange when switch is clicked on', () => {
        // given
        const defaultChecked = true;
        const wrapper = getWrapper({
            defaultChecked
        });
        const customEvent = {
            target: {
                value: values[0].value,
                name: name,
                getAttribute: (attr) => (attr === 'path' ? path : null)
            }
        };
        // when
        wrapper.find('.custom-control-input').simulate('change', { target: { checked: !defaultChecked } });
        // then
        expect(onChange).toHaveBeenCalledWith(expect.objectContaining({
            target: expect.objectContaining({
                value: customEvent.target.value,
                name: customEvent.target.name
            })
        }));
        expect(wrapper.find('.custom-control-label-left').hasClass('active')).toBeTruthy();
    });

    it('should have id when provided', () => {
        // given
        const wrapper = getWrapper({
            id: 'test-id'
        });
        // then
        expect(wrapper.find('#test-id')).toHaveLength(1);
    });

    it('should apply dark theme by default', () => {
        // given
        const wrapper = getWrapper();
        // then
        expect(wrapper.find('.switch--dark')).toHaveLength(1);
    });

    it('should apply light theme', () => {
        // given
        const wrapper = getWrapper({
            theme: 'light'
        });
        // then
        expect(wrapper.find('.switch--light')).toHaveLength(1);
    });
});
