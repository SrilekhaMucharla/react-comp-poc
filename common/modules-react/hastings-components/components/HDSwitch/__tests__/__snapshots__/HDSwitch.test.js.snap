// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HDSwitch should render correctly and match the snapshot 1`] = `
<div
  className="hd-switch switch--dark"
>
  <div
    className="custom-control custom-switch"
  >
    <input
      checked={true}
      className="custom-control-input"
      id="hd-custom-switch"
      name="name"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="custom-control-label-left"
      htmlFor="hd-custom-switch"
    >
      value 1
    </label>
    <label
      className="custom-control-label custom-control-label-right active"
      htmlFor="hd-custom-switch"
    >
      value 2
    </label>
  </div>
</div>
`;
